epochStart/metachain/stakingDataProvider.go

```diff
@@ -257,48 +272,111 @@ func (sdp *stakingDataProvider) GetBlsKeyOwner(blsKey []byte) (string, error) {
+	topUpValue, totalStakedValue, numStakedWaiting, blsKeys, err := sdp.getOwnerInfoFromSC(owner)
 	if err != nil {
 		return nil, err
 	}
 
+	topUpPerNode := big.NewInt(0)
+	numStakedNodes := numStakedWaiting.Int64()
+	if numStakedNodes == 0 {
+		log.Debug("stakingDataProvider.fillOwnerData",
+			"message", epochStart.ErrOwnerHasNoStakedNode,
+			"owner", hex.EncodeToString([]byte(owner)),
+			"validator", hex.EncodeToString(validator.GetPublicKey()),
+		)
+	} else {
+		topUpPerNode = big.NewInt(0).Div(topUpValue, numStakedWaiting)
+	}
+
 	ownerData := &ownerStats{
-		numEligible:        0,
-		numStakedNodes:     numStakedWaiting.Int64(),
-		topUpValue:         topUpValue,
-		totalStaked:        totalStakedValue,
-		eligibleBaseStake:  big.NewInt(0).Set(sdp.minNodePrice),
-		eligibleTopUpStake: big.NewInt(0),
-		topUpPerNode:       big.NewInt(0),
+		numEligible:          0,
+		numStakedNodes:       numStakedNodes,
+		numActiveNodes:       numStakedNodes,
+		totalTopUp:           topUpValue,
+		topUpPerNode:         topUpPerNode,
+		totalStaked:          totalStakedValue,
+		eligibleBaseStake:    big.NewInt(0).Set(sdp.minNodePrice),
+		eligibleTopUpStake:   big.NewInt(0),
+		eligibleTopUpPerNode: big.NewInt(0),
+		qualified:            true,
+	}
+	err = sdp.checkAndFillOwnerValidatorAuctionData([]byte(owner), ownerData, validator)
+	if err != nil {
+		return nil, err
 	}
 
 	ownerData.blsKeys = make([][]byte, len(blsKeys))
 	copy(ownerData.blsKeys, blsKeys)
 
-	sdp.cache[validatorAddress] = ownerData
-
 	return ownerData, nil
 }
 
-func (sdp *stakingDataProvider) getValidatorInfoFromSC(validatorAddress string) (*big.Int, *big.Int, *big.Int, [][]byte, error) {
-	validatorAddressBytes := []byte(validatorAddress)
+func (sdp *stakingDataProvider) checkAndFillOwnerValidatorAuctionData(
+	ownerPubKey []byte,
+	ownerData *ownerStats,
+	validator state.ValidatorInfoHandler,
+) error {
+	validatorInAuction := isInAuction(validator)
+	if !validatorInAuction {
+		return nil
+	}
+	if validatorInAuction && ownerData.numStakedNodes == 0 {
+		return fmt.Errorf("stakingDataProvider.checkAndFillOwnerValidatorAuctionData for validator in auction error: %w, owner: %s, node: %s",
+			epochStart.ErrOwnerHasNoStakedNode,
+			hex.EncodeToString(ownerPubKey),
+			hex.EncodeToString(validator.GetPublicKey()),
+		)
+	}
+	if validatorInAuction && !sdp.flagStakingV4Enable.IsSet() {
+		return fmt.Errorf("stakingDataProvider.checkAndFillOwnerValidatorAuctionData for validator in auction error: %w, owner: %s, node: %s",
+			epochStart.ErrReceivedAuctionValidatorsBeforeStakingV4,
+			hex.EncodeToString(ownerPubKey),
+			hex.EncodeToString(validator.GetPublicKey()),
+		)
+	}
+
+	ownerData.numActiveNodes -= 1
+	ownerData.numAuctionNodes = 1
+	ownerData.auctionList = []state.ValidatorInfoHandler{validator}
+
+	return nil
+}
+
+func (sdp *stakingDataProvider) getOwnerInfoFromSC(owner string) (*big.Int, *big.Int, *big.Int, [][]byte, error) {
```

Comment:
could you add a struct for the return values? There are 3 Ints which could be interchanged.

Reply:
Right, good suggestion.
Added `ownerInfoSC` struct

Reply:
ok